### PR TITLE
[Dubbo-2328]Fix the concurrency limit of 'ActiveLimitFilter' to calculate atomicity

### PR DESCRIPTION
**Pull request by wj1369884616**
## What is the purpose of the change

Solves the problem of 'ActiveLimitFilter' consumer concurrency limit #2328

## Brief changelog

adjust the concurrency limit of 'ActiveLimitFilter' to calculate atomicity by Semaphore

## Verifying this change

org.apache.dubbo.rpc.filter.ActiveLimitFilter
org.apache.dubbo.rpc.RpcStatus

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/incubator-dubbo/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [x] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/incubator-dubbo/tree/master/dubbo-test).
- [x] Run `mvn clean install -DskipTests` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/incubator-dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcStatus.java [hunk 1]
```diff
@@ -53,6 +53,12 @@
     private volatile Semaphore executesLimit;
     private volatile int executesPermits;
 
+    /**
+     * Semaphore used to control concurrency limit set by `actives`
+     */
+    private volatile Semaphore activesLimit;
```
Comment (carryxyh) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcStatus.java]: Is directly init better?
How do u think about it?

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcStatus.java [hunk 2]
```diff
@@ -334,4 +340,28 @@ public Semaphore getSemaphore(int maxThreadNum) {
...
+
+
+    /**
+     * Get the semaphore for thread number. Semaphore's permits is decided by {@link Constants#ACTIVES_KEY}
+     *
+     * @param maxThreadNum value of {@link Constants#ACTIVES_KEY}
+     * @return thread number semaphore
+     */
+    public Semaphore getActivesSemaphore(int maxThreadNum) {
+        if(maxThreadNum <= 0) {
```
Comment (carryxyh) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcStatus.java]: format your code, pls.

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/filter/ActiveLimitFilter.java [hunk 3]
```diff
@@ -37,30 +39,28 @@ public Result invoke(Invoker<?> invoker, Invocation invocation) throws RpcExcept
...
-                                    + invocation.getMethodName() + ", elapsed: " + elapsed
-                                    + ", timeout: " + timeout + ". concurrent invokes: " + active
-                                    + ". max concurrent invoke limit: " + max);
-                        }
-                    }
+            activesLimit = count.getActivesSemaphore(max);
+            try {
+                if(activesLimit != null && !(acquireResult = activesLimit.tryAcquire(timeout,TimeUnit.MILLISECONDS))) {
+                    long elapsed = System.currentTimeMillis() - start;
+                    int active=count.getActive();
```
Comment (carryxyh) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/filter/ActiveLimitFilter.java]: need format, too.

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/filter/ActiveLimitFilter.java [hunk 4]
```diff
@@ -76,8 +76,8 @@ public Result invoke(Invoker<?> invoker, Invocation invocation) throws RpcExcept
             }
         } finally {
             if (max > 0) {
-                synchronized (count) {
-                    count.notify();
+                if(acquireResult){
```
Comment (carryxyh) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/filter/ActiveLimitFilter.java]: format

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcStatus.java (hunk 1 above)
Comment (wj1369884616) [reply dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcStatus.java]: actives may be changed by URL

Comment (wj1369884616) [reply dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcStatus.java]: > Is directly init better?
> How do u think about it?



> Is directly init better?
> How do u think about it?

actives may be changed by url

Comment (wj1369884616) [reply dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcStatus.java]: It may be changed by url

Comment (carryxyh) [reply dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcStatus.java]: Ok, u are right.
Pls deal with your code format.

Comment (wj1369884616) [reply dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcStatus.java]: ok

**Comment by carryxyh** [comment]
It is better to provide some pr for this.

**Comment by kimmking** [comment]
My question is how to test it correctly.

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/filter/ActiveLimitFilter.java [hunk 5]
```diff
@@ -37,30 +39,28 @@ public Result invoke(Invoker<?> invoker, Invocation invocation) throws RpcExcept
...
-                            throw new RpcException("Waiting concurrent invoke timeout in client-side for service:  "
-                                    + invoker.getInterface().getName() + ", method: "
-                                    + invocation.getMethodName() + ", elapsed: " + elapsed
-                                    + ", timeout: " + timeout + ". concurrent invokes: " + active
-                                    + ". max concurrent invoke limit: " + max);
-                        }
-                    }
+            activesLimit = count.getActivesSemaphore(max);
+            try {
+                if(!(acquireResult = activesLimit.tryAcquire(timeout,TimeUnit.MILLISECONDS))) {
```
Comment (kimmking) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/filter/ActiveLimitFilter.java]: how to test it work well or not ?

Comment (wj1369884616) [reply dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/filter/ActiveLimitFilter.java]: what do you think of my test?

**Comment by beiwei30** [comment]
I reconsidered the whole scenario, and will come up with a new solution.

**Pull request merged**